### ISSUE TITLE
Pretraining Divergence

### ISSUE THREAD

**Issue by egoetz** [karpathy/nanoGPT#524]
I have been trying to follow the steps listed under "reproducing GPT-2" from the README.md. Unfortunately, when I run the model, my training always diverges. I have tried switching up my learning rate and gradient accumulation but neither of these tactics seemed to work.

![Screenshot 2024-06-13 144355](https://github.com/karpathy/nanoGPT/assets/36557240/a674b8be-1677-4e79-beee-37cde094eb4a)
(Ignore the false start on the teal line---- I just had to restart the run a few times. Ultimately, that line diverges just like the others)
![image](https://github.com/karpathy/nanoGPT/assets/36557240/6f68da4f-0f51-4c23-b89b-2cbedb5e71d0)
(Here the yellow/orange line has a gradient accumulation of 20)

Here are some of my meta-parameters:
batch_size = 24
block_size = 1024
max_iters = 300000
lr_decay_iters = 300000
eval_interval = 1000
eval_iters = 200
log_interval = 100
weight_decay = 5e-2

Note that my learning rate decay iteration is half that of the repo's suggested value. This is because I was trying to do a two-stage training process split into the first 300000 iterations and the second 300000 iterations. Could that be causing an issue? I assumed because I made the decay interval shorter, if anything that would just result in stalled progress, not divergence. I am running this model on 4 A100 80GB GPUs.

**Comment by iminfine** [comment]
This is caused by flash attention. Please disable it and use original self attention.

And also use the default training config, here is batch setting for 4 A100 80g:  

30 batch size * 1024 block size * 4 gradaccum * 4 GPUs = 491,520

**Comment by egoetz** [comment]
How do you disable flash attention? I can't find anything on the torch website which suggests it is togglable.

**Comment by egoetz** [comment]
Is there a way to find the correct configuration for an arbitrary setup? I'd also like to know the suggested parameters for 2 nodes that have 4 GPUs each.

**Comment by melisilaydabal** [comment]
@egoetz I also observe loss explosion when flash attention is enabled. How did changing the dtype help with this problem?

**Comment by MarkoKarbevski** [comment]
Your weight decay is relatively high as well. Where did you find the 5e-2 from? That is 0.05 < 0.1 which is the "original".

**Comment by abs-hello** [comment]
> [@egoetz](https://github.com/egoetz) I also observe loss explosion when flash attention is enabled. How did changing the dtype help with this problem?

Hi, did you change any parameters? I want to reproduce the bug for numerical errors analysis of flash attention. Please give me some details. Thanks!
